docs/KernelFeatures/UsingSecureVault.md

```diff
@@ -0,0 +1,108 @@
+2. Master Key Reader
+   The default implementation of MasterKeyReader gets a list of required passwords from the SecretRepository and provides the values for those passwords by reading system properties, environment variables and master-keys.yaml file.
+
+## How To Use SecureVault
+SecureVault is by default enabled. It reads the aliases and passwords given in the secrets.properties file. The secrets.properties file may contain both plain text and encrypted passwords. We have a separate tool called "ciphertool" (shipped with Carbon /bin) to encrypt the secrets.properties file. Once the tool is run, it will encrypt all the plaintext password in the secrets.properties file.
+
+CipherTool also depends on the configurations given in the secure-vault.yaml. Hence making necessary changes in the "secure-vault.yaml" file before running the CipherTool is mandatory. Once configured, running the "ciphertool" is simple as running the "ciphertool.sh" on Linux and Mac and running the "ciphertool.bat" on Windows machines.
+
+## How To Implement Custom Master Key Reader
+MasterKeyReader (org.wso2.carbon.kernel.securevault.MasterKeyReader) is the interface which all the MasterKeyReader implementations should derive from and should be registered as an OSGi service. SecureVault gets all the MasterKeyReader implementations and binds itself only with the matching MasterKeyReader which is specified in the "secure-vault.yaml" file.
+
+    public class CustomHardCodedMasterKeyReader implements MasterKeyReader {
+        private static Logger logger = LoggerFactory.getLogger(DefaultHardCodedMasterKeyReader.class);
+
+        @Override
+        public void init(MasterKeyReaderConfiguration masterKeyReaderConfiguration) throws SecureVaultException {
+            // No initializations needed for the DefaultMasterKeyReader
+        }
+
+        @Override
+        public void readMasterKeys(List<MasterKey> masterKeys) throws SecureVaultException {
+            logger.debug("Providing hard coded secrets for 'keyStorePassword' and 'privateKeyPassword'");
+
+            MasterKey keyStorePassword = SecureVaultUtils.getSecret(masterKeys, JKSBasedCipherProvider.KEY_STORE_PASSWORD);
+            keyStorePassword.setMasterKeyValue("wso2carbon".toCharArray());
+
+            MasterKey privateKeyPassword = SecureVaultUtils.getSecret(masterKeys, JKSBasedCipherProvider.PRIVATE_KEY_PASSWORD);
+            privateKeyPassword.setMasterKeyValue("wso2carbon".toCharArray());
+        }
+    }
+
+## How To Implement Secret Repository
+All the SecretRepository implementations should derive from the interface SecretRepository and register itself as an OSGi service of that interface. From all the registered implementations for SecretRepository, SecureVault chooses the correct SecretRepository based on the configurations given in the secure-vault.yaml file.
```

Comment:
All the SecretRepository implementations should derive from the SecretRepository interface and should be registered as an OSGi service of that interface.